File: data-structure/graph/mst/mst_kruskal.py
def make_set(x):
    p[x] = x


def find_set(x):
    if p[x] == x:
        return x
    p[x] = find_set(p[x])
    return p[x]


def union(x, y):
    px = find_set(x)
    py = find_set(y)
    # 높이 기준으로 합치는 방식으로 진행
    if rank[px] > rank[py]:
        p[py] = px
    else:
        p[px] = py
        if rank[px] == rank[py]:
            rank[py] += 1


# 가장 비용이 작은 간선부터 선택해나간다
V, E = map(int, input().split())
adj = []
for i in range(E):
    s, e, v = list(map(int, input().split()))
    adj.append([s, e, v])
adj.sort(key=lambda x: x[2])


p = [0] * V
for i in range(V):
    make_set(i)
rank = [0] * V

mst = []
cnt = 0
result = 0
for i in range(E):
    st, en, va = adj[i][0], adj[i][1], adj[i][2]
    if find_set(st) == find_set(en):
        continue
    result += va
    mst.append(adj[i])
    union(st, en)
    cnt += 1
    if cnt == V - 1:
        break


print(p)
print(rank)
print(mst)
print(result)

# 참고
# https://m.blog.naver.com/ssarang8649/221038259400

File: data-structure/graph/mst/MstKruskal.java
import java.util.Arrays;
import java.util.PriorityQueue;
import java.util.stream.IntStream;

public class Main {
    public static void main(String[] args){
        Mst mst = new Mst();
        int minLength = mst.kruskal(
                7,
                11,
                new int[][]{
                        {0, 5, 60},
                        {0, 1, 32},
                        {0, 2, 31},
                        {0, 6, 51},
                        {1, 2, 21},
                        {2, 4, 46},
                        {2, 6, 25},
                        {3, 4, 34},
                        {3, 5, 18},
                        {4, 5, 40},
                        {4, 6, 51}
                }
        );
        System.out.println(minLength);
    }
}

class Mst {
    public int kruskal(int v, int e, int[][] edges) {
        PriorityQueue<Edge> pq = new PriorityQueue<>();
        Arrays
            .stream(edges)
            .forEach((int[] edge) -> pq.add(new Edge(edge[0], edge[1], edge[2])));

        int minLength = 0;
        DisjointSet disjointSet = new DisjointSet(v);

        while (!pq.isEmpty()) {
            Edge edge = pq.poll();

            int px = disjointSet.findSet(edge.start);
            int py = disjointSet.findSet(edge.end);

            if (px == py) {
                continue;
            }

            minLength += edge.distance;
            disjointSet.unionSet(px, py);
        }

        return minLength;
    }
}

class DisjointSet {
    int size;
    int[] parent;

    DisjointSet(int size) {
        this.size = size;
        initParent(size);
        IntStream.range(0, size).forEach(this::makeSet);
    }

    void initParent(int size) {
        parent = new int[size];
    }

    void makeSet(int x) {
        parent[x] = x;
    }

    int findSet(int x) {
        if (parent[x] != x) {
            parent[x] = findSet(parent[x]);
        }
        return parent[x];
    }

    void unionSet(int x, int y) {
        int px = findSet(x);
        int py = findSet(y);

        if (px < py) {
            parent[py] = px;
        } else {
            parent[px] = py;
        }
    }
}

class Edge implements Comparable<Edge> {
    int start;
    int end;
    int distance;

    Edge(int start, int end, int distance) {
        this.start = start;
        this.end = end;
        this.distance = distance;
    }

    @Override
    public int compareTo(Edge edge) {
        return this.distance - edge.distance;
    }
}

// 참고
// https://blog.naver.com/ssarang8649/221038259400
// https://cjh5414.github.io/priorityqueue/
